### PR TITLE
CKEDITOR-434: Allow to upload attachment using temporary upload endpoint 

### PR DESCRIPTION
**Pull request by surli**
* Check if the editor support the new temporary attachment upload and
    use it if it's the case
  * Automatically inject new inputs whenever a file is uploaded when the
    feature is used
  - [x] test on older versions of XWiki to ensure it doesn't break old upload feature
  - [x] test upload size limitation

### CODE REVIEW

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 1]
```diff
@@ -375,6 +375,12 @@ define('xwiki-ckeditor', [
         }))
       };
     }
+    // Define if the temporary attachment feature can be used.
+    if ($(element).attr('data-isTemporaryAttachmentSupported') === 'true') {
+      config['xwiki-upload'] = {
+        isTemporaryAttachmentSupported: true
+      };
```
Comment (manuelleduc) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:382]: Isn't there a risk that other `xwiki-upload` configuration properties get overridden?

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:382]: Unless I'm missing something, but there's no xwiki-upload configuration right now.

Comment (manuelleduc) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:382]: I agree, it could just make the life of future maintainers easier to assume that a config could exist (I'm nitpicking tbh).

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:382]: Note sure how that's even possible: the config object is defined just few lines above (https://github.com/xwiki-contrib/application-ckeditor/blob/52971c8c499bac38920343a5fb1f3bd24461da06/application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml#L307), without any xwiki-upload yet. So if any other xwiki-upload config, they would be added afterwards.

Comment (manuelleduc) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:382]: Indeed, I was reading the code as if it was in a plugin. My bad.

**Comment by surli** [comment]
Note that this PR depends on https://github.com/xwiki/xwiki-platform/pull/1815

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml [hunk 2]
```diff
@@ -97,6 +98,25 @@
...
+    #set ($document = $doc)
+    #if ("$!request.document" != '')
+      #set ($document = $xwiki.getDocument($request.document))
+    #end
+    #set ($reference = $document.documentReference)
+    #set ($attachment = $services.wysiwyg.temporaryUploadAttachment($reference, 'upload'))
+    #if ($attachment)
+      #sendSuccess($document, $attachment.filename)
+    #else
+      #sendError(400 "The 'upload' request parameter is missing. Either you didn't submit any file or the size of the submitted file exceeds the configured maximum upload size.")
```
Comment (surli) [inline application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:113]: This actually needs to be tested with a maximum upload size since we now throw an exception.

File: application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml [hunk 3]
```diff
@@ -214,12 +214,14 @@
...
     "$!services.wysiwyg.importOfficeAttachment($missingAttachmentReference, {'useOfficeViewer': true})" != '')
+  #set ($temporaryAttachmentSupported = $services.wysiwyg.isTemporaryAttachmentSupported())
   #set ($discard = $parameters.attributes.putAll({
     'class': "$!parameters.attributes.get('class') ckeditor-textarea loading",
     'spellcheck': false,
     'data-syntax': ${services.wysiwyg.HTMLSyntax.toIdString()|'annotatedxhtml/1.0'},
     'data-sourceDocumentSyntax': $sourceDocument.syntax.toIdString(),
-    'data-officeImporterSupported': $officeImporterSupported
+    'data-officeImporterSupported': $officeImporterSupported,
+    'data-isTemporaryAttachmentSupported': true
```
Comment (surli) [inline application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml:224]: That's actually not good as the feature is not there in older version of XWiki: we need to put this value based on a check of the upload API endpoint.

File: application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml [hunk 4]
```diff
@@ -161,6 +161,7 @@
       'data-sourceDocumentReference': XWiki.Model.serialize(config.document.documentReference),
       'data-sourceDocumentSyntax': config.document.syntax,
       'data-officeImporterSupported': config.enableOfficeImport,
+      'data-isTemporaryAttachmentSupported': true,
```
Comment (surli) [inline application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml:164]: This needs to be changed too.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 5]
```diff
@@ -390,6 +400,29 @@ define('xwiki-ckeditor', [
...
   };
 
+  // Define a way to switch on or off the temporary attachment upload mechanism.
+  ckeditor.switchTemporaryAttachmentUpload = function (isEnabled, instance, config) {
+    if (!config) {
+      var config = instance.config;
+    }
+    if (!config['xwiki-upload']) {
+      config['xwiki-upload'] = {
+        'isTemporaryAttachmentSupported': true
```
Comment (surli) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:410]: This is not good either.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 6]
```diff
@@ -390,6 +400,29 @@ define('xwiki-ckeditor', [
     return oldInline.call(this, element, ckeditor.tools.extend(getConfig(element), config, true));
   };
 
+  // Define a way to switch on or off the temporary attachment upload mechanism.
+  ckeditor.switchTemporaryAttachmentUpload = function (isEnabled, instance, config) {
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: I don't understand the need for this. If someone wants to disable the temporary attachment upload for a specific CKEditor instance they can do it:
* either by passing an instance configuration object when creating the editor, if they control the editor creation
```
ckeditor.replace(textarea, {
  'xwiki-upload': {
    'isTemporaryAttachmentSupported': false
  }
});
```
* or by listening to CKEditor events as indicated on https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor+Integration#HConfiguretheEditor and changing the config after it is loaded.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 7]
```diff
@@ -390,6 +400,29 @@ define('xwiki-ckeditor', [
...
+    }
+
+    if (isEnabled !== config['xwiki-upload'].isTemporaryAttachmentSupported) {
+      var sourceDocument = getDocument(element);
+      config.filebrowserUploadUrl = getUploadURL(sourceDocument, 'filebrowser', isEnabled);
+      config.uploadUrl = getUploadURL(sourceDocument, 'filetools', isEnabled);
+      config['xwiki-upload'].isTemporaryAttachmentSupported = isEnabled;
+      $(element).attr('data-isTemporaryAttachmentSupported', isEnabled);
+      // Those properties should never be overridden by the in-page configuration.
+      ckeditor.keepConfigValues = ['xwiki-upload', 'filebrowserUploadUrl', 'uploadUrl'];
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:422]: Why? I don't see why they shouldn't. This looks hackish.

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml [hunk 8]
```diff
@@ -89,11 +90,30 @@
       #end
       #createAttachment($fileName)
     #else
-      #sendError(403 'The CSRF token is missing.')
+      #sendError(403 "$services.localization.render('ckeditor.upload.error.csrf')")
     #end
   #else
     ## Bad Request
-    #sendError(400 "The 'upload' request parameter is missing. Either you didn't submit any file or the size of the submitted file exceeds the configured maximum upload size.")
+    #sendError(400 "$services.localization.render('ckeditor.upload.error.emptyreturn')")
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:97]:
```suggestion
    #sendError(400 "$services.localization.render('ckeditor.upload.error.emptyReturn')")
```

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml [hunk 9]
```diff
@@ -89,11 +90,30 @@
...
+
+## New way of handling attachment without saving immediately the document.
+#macro (handleTemporaryUploadRequest)
+  #if ($services.csrf.isTokenValid($request.form_token))
+    #set ($document = $doc)
+    #if ("$!request.document" != '')
+      #set ($document = $xwiki.getDocument($request.document))
+    #end
+    #set ($reference = $document.documentReference)
+    #set ($attachment = $services.temporaryattachments.uploadTemporaryAttachment($reference, 'upload'))
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:109]: Why lowercase ``temporaryattachments``? Why not camelcase?

```suggestion
    #set ($attachment = $services.temporaryAttachments.uploadTemporaryAttachment($reference, 'upload'))
```

File: application-ckeditor-ui/src/main/resources/CKEditor/Translations.xml [hunk 10]
```diff
@@ -116,6 +116,9 @@ ckeditor.plugin.syntax.wysiwygModeUnsupported=The WYSIWYG edit mode is not suppo
 
 ckeditor.inline.failedToInit=Failed to initialize CKEditor.
 
+ckeditor.upload.error.csrf=The CSRF token is missing.
+ckeditor.upload.error.emptyreturn=Error when uploading the file: it might exceed the configured maximum upload size.
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/Translations.xml:120]:
```suggestion
ckeditor.upload.error.emptyReturn=Error when uploading the file: it might exceed the configured maximum upload size.
```

File: application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml [hunk 11]
```diff
@@ -214,12 +214,14 @@
...
     "$!services.wysiwyg.importOfficeAttachment($missingAttachmentReference, {'useOfficeViewer': true})" != '')
+  #set ($temporaryAttachmentSupported = ("$!services.temporaryattachments" != ""))
   #set ($discard = $parameters.attributes.putAll({
     'class': "$!parameters.attributes.get('class') ckeditor-textarea loading",
     'spellcheck': false,
     'data-syntax': ${services.wysiwyg.HTMLSyntax.toIdString()|'annotatedxhtml/1.0'},
     'data-sourceDocumentSyntax': $sourceDocument.syntax.toIdString(),
-    'data-officeImporterSupported': $officeImporterSupported
+    'data-officeImporterSupported': $officeImporterSupported,
+    'data-isTemporaryAttachmentSupported': $temporaryAttachmentSupported
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml:224]:
```suggestion
    'data-temporary-attachment-supported': $temporaryAttachmentSupported
```

I know I used camelcase above, but it's not right.

File: application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml [hunk 12]
```diff
@@ -214,12 +214,14 @@
   #end
   #set ($officeImporterSupported = $officeServerConnected &amp;&amp;
     "$!services.wysiwyg.importOfficeAttachment($missingAttachmentReference, {'useOfficeViewer': true})" != '')
+  #set ($temporaryAttachmentSupported = ("$!services.temporaryattachments" != ""))
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml:217]:
```suggestion
  #set ($temporaryAttachmentSupported = "$!services.temporaryAttachments" != '')
```

File: application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml [hunk 13]
```diff
@@ -161,6 +161,7 @@
       'data-sourceDocumentReference': XWiki.Model.serialize(config.document.documentReference),
       'data-sourceDocumentSyntax': config.document.syntax,
       'data-officeImporterSupported': config.enableOfficeImport,
+      'data-isTemporaryAttachmentSupported': (config.enableTemporaryAttachmentUpload === true),
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml:164]:
```suggestion
      'data-temporary-attachment-supported': config.enableTemporaryAttachmentUpload === true,
```

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 14]
```diff
@@ -304,16 +310,17 @@ define('xwiki-ckeditor', [
       }
     }
 
+    var isTemporaryAttachmentSupported = $(element).attr('data-isTemporaryAttachmentSupported') || false;
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:313]:
```suggestion
    var temporaryAttachmentSupported = $(element).attr('data-temporary-attachment-supported') || false;
```

Attributes should use lowercase. Use dash to separate tokens.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 15]
```diff
@@ -268,24 +268,30 @@ define('xwiki-ckeditor', [
 
   // We have to pass the plugin that makes the request (the initiator) because the expected response can be different
   // (e.g. between the filebrowser and filetools plugins).
-  var getUploadURL = function(document, initiator) {
+  var getUploadURL = function(document, initiator, allowTemporaryAttachments) {
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:271]: Adding the flag to the upload URL **here** means the URL needs to be updated when the configuration changes. This complicates a lot the code (I really don't like ``switchTemporaryAttachmentUpload``). I prefer to modify the ``xwiki-upload`` plugin to listen to https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-fileUploadRequest and set a custom HTTP header that can be read from the ``FileUploader``.

**Review by mflorea — CHANGES_REQUESTED** [review]

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 6 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: I wanted to provide the capability of disabling the feature even through the advanced WYSIWYG configuration in XWiki, as a failsafe mechanism to disable it globally if needed. AFAIU the options you present here, wouldn't work there. 
Now maybe it was a bit of overengineering to provide this: we can always document that users can edit EditSheet for example to disable it, or any other sheet including the editor. Or programmatically like you described.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 7 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:422]: This is a bit hackish, it was to make working the disabling of the feature in the advanced editor configuration in XWiki. But as I said in a previous comment, maybe that's not needed.

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml (hunk 9 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:109]: Hmmm, sounds like we have a mix of camelcase and lowercase script services, and I'm more used to use the lowercase ones, so I always use lowercase, e.g:
  * $services.officeimporter
  * $services.modelvalidation
  * $services.linkchecker
 
The only two script service using camelcase I can see are:
  * $services.liveData
  * $services.messageStream

Not sure we have a rule for this.

File: application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml (hunk 13 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml:164]: Need to fix this also in xwiki side.

File: application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js [hunk 16]
```diff
@@ -139,4 +142,16 @@
     //   see https://ckeditor.com/docs/ckeditor4/latest/guide/dev_file_upload.html#request-2
     }, null, null, 1);
   };
+
+  // Inject a new input field when an attachment is added so that the save request knows which are the new attachments.
+  var listenUploadedAttachments = function(editor) {
```
Comment (surli) [inline application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js:147]: Need to fix the method name too.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 15 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:271]: > means the URL needs to be updated when the configuration changes.

Do we agree that if I remove `switchTemporaryAttachmentUpload` and rely entirely on the configuration flag, as you proposed in https://github.com/xwiki-contrib/application-ckeditor/pull/50/files#r848825407 we don't need to support updating it since the call to replace would do it? I'm not sure here.

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml (hunk 9 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:109]: Ok tbh I don't have a strong opinion on this, so I will change if you really prefer. In the meantime, I'm opening a discussion thread on the forum to have a more clear rule for this in the future.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 15 above)
Comment (mflorea) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:271]: The upload URL would be the same no matter if temporary attachments are allowed / supported or not. The ``xwiki-upload`` plugin will listen to the upload request and pas the information to the server to indicate if temporary attachments are allowed / supported. The easiest way to pass this information is through some custom HTTP headers I think. See for instance https://ckeditor.com/docs/ckeditor4/latest/guide/dev_file_upload.html#request-2 .

> You can also add custom request headers or set flags for the default request.

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:271]: So if I understand properly what you're saying, the idea is that on each `fileUploadRequest` to check the editor config and set a request header with the `allowTemporaryAttachments` flag value, then in FileUploaders use that header value (if present) to decide which upload to use. Is that correct?

File: application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml (hunk 9 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/FileUploader.xml:109]: for info: https://forum.xwiki.org/t/rules-for-script-service-naming/10138

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 15 above)
Comment (mflorea) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:271]: Yes

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 6 above)
Comment (mflorea) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: If we want to allow users to disable the usage of temporary attachments in CKEditor globally, for all editor instances, then we need to make this configuration global. Currently this configuration is implemented as instance level configuration: the value is read from the HTML element that is enhanced by the CKEditor (``data-temporary-attachment-supported``), so two elements can use different values, thus is cannot be overwritten globally. In order to make it a global config we need to:

* stop reading it from the editor element
* instead, read it inside ``CKEditor.Config`` (advanced config text area) but:
  * ``CKEditor.Config`` doesn't exist by default, so we'd have to provide a default one (either in CKEditor Application or in XS)
  * the advanced config text area doesn't support Velocity code ATM, we'd have to fix this (i.e. https://jira.xwiki.org/browse/XWIKI-12301) if we want to have the default ``CKEditor.Config`` in the CKEditor Application (it needs to detect the support for temporary attachments using Velocity code)
    * alternatively, we could make XS provide a default ``CKEditor.Config`` page, with configuration specific to XS, and this configuration would have in the advanced text area:
      ```
      config['xwiki-upload'].temporaryAttachmentSupported = true;
      ```
  * the Reset action would have to be modified to restore the version provided by the extension (in case ``CKEditor.Config`` is part of an extension, e.g. XS flavor) instead of simply deleting the configuration page

I think the cleanest, on the long term at lest, is to:

* make XS (flavor) provide a default ``CKEditor.Config`` page with configuration specific to XS
  * set ``config['xwiki-upload'].temporaryAttachmentSupported = true;`` in its advanced configuration text area
  * modify the Reset action from ``AdminSection`` to restore the version from the extension rather than deleting the ``CKEditor.Config`` page
* no need to modify ``EditSheet``, ``VelocityMacros`` or ``InlineEditor``
* ``xwiki-upload`` reads this configuration and sets a custom header on each upload request, also listeners to upload response to add the hidden input
* ``FileUploader`` checks the custom HTTP header to know what type of upload to do
* anyone that needs to disable temporary attachment usage can either do this from the instance config, if they control the editor creation, or by listening to editor events and modifying the config just before the editor is fully loaded, as per https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor+Integration#HConfiguretheEditor .

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: Thanks for the heads up, I won't go there right now: I'm not even sure that there's a need to disable it globally right now, and as you said we have several mechanism available to disable it where it's needed.

Comment (mflorea) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: Note that an intermediary step in this direction, that removes the need to provide a default ``CKEditor.Config`` and to modify the Reset action, but still doesn't require any changes to ``EditSheet``, ``VelocityMacros`` or ``InlineEditor`` is to modify ``ConfigSheet``, precisely here https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-ui/src/main/resources/CKEditor/ConfigSheet.xml#L59 . We can make this the place to put global config that needs some Velocity code for evaluation. Administrators will still be able to disable temporary attachment easily from the advanced config text area, because that is executed, in JavaScript, on top of the custom config we set in ``ConfigSheet``. Basically:

```
#set ($temporaryAttachmentSupported = "$!services.temporaryAttachments" != '')
## Global default configuration that doesn't depend on the edited content and that needs Velocity code for evaluation.
## Static global configuration (that doesn't need Velocity) should be put in config.js . The default instance level
## configuration is in EditSheet.
#set ($customConfig = {
  'xwiki-upload': {
    'temporaryAttachmentSupported': $temporaryAttachmentSupported
  }
})
```

This means that we'll have this levels of configuration:

* ``config.js``: global **default** (static) configuration, no Velocity evaluation
* ``CKEditor.ConfigSheet``: global **default** (dynamic) configuration, with Velocity evaluation
* ``CKEditor.Config``: global (static) configuration, no Velocity evaluation
* ``CKEditor.EditSheet``: instance **default** (dynamic) configuration
* instance configuration passed when creating the editor
* custom instance configuration done by listening to editor events (from code that doesn't control the editor instance creation)

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: The idea sounds interesting but there's one thing I'm not sure about: it means that users might customize the `CKEditor.ConfigSheet` document, which is not really aimed to be customized, so in case of upgrade it might be problematic no?

Comment (mflorea) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: No, ``CKEditor.ConfigSheet`` is not meant to be customized by users. It provides **default** configuration which can be overwritten from the administration section. I documented this on https://extensions.xwiki.org/xwiki/bin/view/Extension/CKEditor%20Integration/#HConfigurationLevels .

Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:404]: Right, so sounds good then.

File: application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js [hunk 17]
```diff
@@ -26,6 +26,8 @@
 
     init: function(editor) {
       preventParallelUploads(editor);
+      listenToUploadedAttachments(editor);
+      listenUploadRequest(editor);
```
Comment (mflorea) [inline application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js:30]: Why do you need two functions? :) for me they are related. I'd keep a single ``listenToUploadedAttachments``.

File: application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js [hunk 18]
```diff
@@ -139,4 +141,28 @@
...
+        xhr.setRequestHeader( 'X-XWiki-Temporary-Attachment-Support',
+          editor.config['xwiki-upload'].isTemporaryAttachmentSupported);
+      }
+    });
+  };
+
+  // Inject a new input field when an attachment is added so that the save request knows which are the new attachments.
+  var listenToUploadedAttachments = function(editor) {
+    editor.on('fileUploadResponse', function (event) {
+      if ((editor.config['xwiki-upload'] || {}).isTemporaryAttachmentSupported) {
```
Comment (mflorea) [inline application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js:158]: You don't do this fail-safe above. Either it's not needed and thus you can remove it or it's needed and you need to do it above also. BTW, I think you're missing https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-plugins/src/main/resources/xwiki-macro/plugin.js#L24 for the 'xwiki-upload' config.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml [hunk 19]
```diff
@@ -304,6 +309,7 @@ define('xwiki-ckeditor', [
       }
     }
 
+    var isTemporaryAttachmentSupported = $(element).attr('data-temporary-attachment-supported') || false;
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:312]: Do you still need this? I don't think so, since the config is now done in ``ConfigSheet``. No change is required to ``EditSheet`` I think.

File: application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml [hunk 20]
```diff
@@ -161,6 +161,7 @@
       'data-sourceDocumentReference': XWiki.Model.serialize(config.document.documentReference),
       'data-sourceDocumentSyntax': config.document.syntax,
       'data-officeImporterSupported': config.enableOfficeImport,
+      'data-temporary-attachment-supported': config.enableTemporaryAttachmentUpload === true,
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/InlineEditor.xml:164]: I don't think this is needed anymore now that the config is done in ``ConfigSheet``.

File: application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml [hunk 21]
```diff
@@ -214,12 +214,14 @@
   #end
   #set ($officeImporterSupported = $officeServerConnected &amp;&amp;
     "$!services.wysiwyg.importOfficeAttachment($missingAttachmentReference, {'useOfficeViewer': true})" != '')
+  #set ($temporaryAttachmentSupported = "$!services.temporaryAttachments" != '')
```
Comment (mflorea) [inline application-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml:217]: I don't think this is needed anymore now that the config is done in ``ConfigSheet``.

**Review by mflorea — CHANGES_REQUESTED** [review]

File: application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js (hunk 17 above)
Comment (surli) [reply application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js:30]: Ok I'll group them.

File: application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js (hunk 18 above)
Comment (surli) [reply application-ckeditor-plugins/src/main/resources/xwiki-upload/plugin.js:158]: >  BTW, I think you're missing https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-plugins/src/main/resources/xwiki-macro/plugin.js#L24 for the 'xwiki-upload' config.

Right. 

> it's needed and you need to do it above also

Not sure it's needed anymore, since the upgrade of CKEditor will provide it. So I guess it's safe without it now.

File: application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml (hunk 19 above)
Comment (surli) [reply application-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:312]: Right

**Review by mflorea — APPROVED** [review]
